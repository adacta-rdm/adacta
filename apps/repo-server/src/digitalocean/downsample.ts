import { downsampleHTTPEndpoint } from "../microservices/downsampleHTTPEndpoint";

import { createDigitalOceanFunction } from "~/lib/serverless/digitalocean/createDigitalOceanFunction";

export const main = createDigitalOceanFunction(downsampleHTTPEndpoint);
