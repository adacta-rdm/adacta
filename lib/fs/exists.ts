import { access } from "fs/promises";

import type { PathArg } from "./private/normalizePath";
import { normalizePath } from "./private/normalizePath";

export async function exists(path: PathArg) {
	try {
		await access(normalizePath(path));
		return true;
	} catch (e) {
		return false;
	}
}
