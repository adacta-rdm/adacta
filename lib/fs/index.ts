/**
 * @file Automatically generated by barrelsby.
 */

export * from "./exists";
export * from "./isDir";
export * from "./mkdirTmp";
export * from "./mkdirp";
export * from "./readFile";
export * from "./readJSONFile";
export * from "./readUTF8File";
export * from "./readdir";
export * from "./rmrf";
export * from "./stat";
export * from "./writeFile";
export * from "./private/normalizePath";
