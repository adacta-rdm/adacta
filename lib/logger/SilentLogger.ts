import { PassThrough } from "stream";

import { LOG_LEVEL, Logger } from "./Logger";

/* eslint-disable class-methods-use-this */

/**
 * A logger that does not log anything.
 */
export class SilentLogger extends Logger {
	constructor() {
		super({ level: LOG_LEVEL.SILENT, stream: new PassThrough() });
	}

	trace(): void {}

	debug(): void {}

	info(): void {}

	error(): void {}

	warn(): void {}
}
