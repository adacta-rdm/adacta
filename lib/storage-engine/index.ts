/**
 * @file Automatically generated by barrelsby.
 */

export * from "./FileNotFoundError";
export * from "./FileSystemStorageEngine";
export * from "./ReusableFileHandle";
export * from "./S3StorageEngine";
export * from "./StorageEngine";
export * from "./refreshableTimeout";
