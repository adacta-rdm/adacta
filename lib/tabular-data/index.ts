/**
 * @file Automatically generated by barrelsby.
 */

export * from "./TabularData";
export * from "./interface/StrictArrayBuffer";
