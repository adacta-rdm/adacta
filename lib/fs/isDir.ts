import { stat } from "fs/promises";

import type { PathArg } from "./private/normalizePath";
import { normalizePath } from "./private/normalizePath";

export async function isDir(path: PathArg) {
	try {
		const stats = await stat(normalizePath(path));
		return stats.isDirectory();
	} catch (e: any) {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
		if (e.code && e.code === "ENOENT") {
			return false;
		}
		throw e;
	}
}
