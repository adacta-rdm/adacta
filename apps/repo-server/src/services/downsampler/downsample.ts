import type { StorageEngine } from "@omegadot/storage-engine";
import { TabularData } from "@omegadot/tabular-data";

import type { Point } from "./downsampleLTTBRowMajorAsync";
import { downsampleLTTBRowMajorAsync } from "./downsampleLTTBRowMajorAsync";
import type { IDownsamplingTaskArgs } from "../../../interface/IDownsamplingTaskArgs";

import type { Logger } from "~/lib/logger/Logger";

/**
 * Produces downsampled data from a given TabularData input.
 *
 * Note that the number of columns in the output tables will likely differ from those of the input file. This is because
 * a new set of x values is generated for every y column. The number of columns in the output TabularData is twice the
 * number of y-columns in the input.
 *
 * For example, from a TabularData resource with a single x-column that is common to the remaining 3 y-columns,
 * ┌─────┬─────┬─────┬─────┐
 * │x    │y1   │y2   │y3   │
 * ├─────┼─────┼─────┼─────┤
 * │     │     │     │     │ 1
 * │     │     │     │     │
 * │     │     │     │     │ ...
 * │     │     │     │     │
 * │     │     │     │     │ N (`input.numberRows`)
 * └─────┴─────┴─────┴─────┘
 *
 * the function will create TabularData files with the following structure:
 * ┌─────┬─────┬─────┬─────┬─────┬─────┐
 * │x1   │y1   │x2   │y2   │x3   │y3   │
 * ├─────┼─────┼─────┼─────┼─────┼─────┤
 * │     │     │     │     │     │     │ 1
 * │     │     │     │     │     │     │ ...
 * │     │     │     │     │     │     │ N (`output.threshold`)
 * └─────┴─────┴─────┴─────┴─────┴─────┘
 *
 * Every downsampled TabularData will have the same columns, but a different number of rows (matching the entries under
 * `output.threshold`).
 */
export async function downsample(
	ctx: { sto: StorageEngine; logger: Logger },
	args: IDownsamplingTaskArgs
): Promise<Point[][]> {
	const { input, threshold } = args;

	ctx.logger.info(
		`Downsampling input ${input.path} with ${input.numberRows} data points to ${threshold} data points.`
	);

	const stream = TabularData.createReadStream(ctx.sto, input.path, input.numberColumns);

	return downsampleLTTBRowMajorAsync(
		stream,
		input.numberRows,
		threshold,
		input.columns.y.map((y) => ({ x: input.columns.x, y }))
	);
}
