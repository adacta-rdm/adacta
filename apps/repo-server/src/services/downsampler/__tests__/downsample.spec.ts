import { mkdirTmp } from "@omegadot/fs";
import { FileSystemStorageEngine } from "@omegadot/storage-engine";
import type { Writable } from "@omegadot/streams";
import { TabularData } from "@omegadot/tabular-data";
import { describe, test, expect } from "vitest";

import { downsample } from "../downsample";

import { SilentLogger } from "~/lib/logger/SilentLogger";

function fillTable(stream: Writable<number[]>) {
	stream.write([0, 2, 9]);
	stream.write([1, 4, 1]);
	stream.write([2, 6, 1]);
	stream.write([3, 8, 1]);
	stream.write([4, 2, 9]);
	stream.write([5, 4, 5]);
	stream.write([6, 6, 9]);
	stream.write([7, 8, 9]);
	stream.write([8, 2, 1]);
	stream.write([9, 4, 5]);
	stream.end();

	return stream.promise();
}

describe("downsample()", () => {
	test("downsample all columns in tabular data", async () => {
		const sto = new FileSystemStorageEngine(await mkdirTmp());

		await fillTable(TabularData.createWriteStream(sto, "temp"));

		const result = await downsample(
			{ sto, logger: new SilentLogger() },
			{
				input: {
					prefix: "",
					path: "temp",
					numberRows: 10,
					numberColumns: 3,
					columns: {
						x: 0,
						y: [1, 2],
					},
				},
				threshold: 5,
			}
		);

		// Returns an a
		expect(result).toHaveLength(2);
		// 5 points in y1
		expect(result[0]).toHaveLength(5);
		// 5 points in y2
		expect(result[1]).toHaveLength(5);

		expect(result).toMatchSnapshot();
	});
});
