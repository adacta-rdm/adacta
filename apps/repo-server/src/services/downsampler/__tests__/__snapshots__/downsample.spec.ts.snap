// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`downsample() > downsample all columns in tabular data 1`] = `
[
  [
    [
      0,
      2,
    ],
    [
      2,
      6,
    ],
    [
      4,
      2,
    ],
    [
      7,
      8,
    ],
    [
      9,
      4,
    ],
  ],
  [
    [
      0,
      9,
    ],
    [
      1,
      1,
    ],
    [
      4,
      9,
    ],
    [
      8,
      1,
    ],
    [
      9,
      5,
    ],
  ],
]
`;
